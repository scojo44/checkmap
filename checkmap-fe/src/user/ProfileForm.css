.ProfileForm ul {
  list-style-type: none;
}
.ProfileForm .list-row {
  display: flex;
  justify-content: space-between;
  margin-bottom: .5rem;
}
.ProfileForm .list-name {
  flex: 2;
}
.ProfileForm .list-button {
  flex: 1;
}
.ProfileForm input[name="imageURL"] {
  display: block;
  width: 95%;
}
