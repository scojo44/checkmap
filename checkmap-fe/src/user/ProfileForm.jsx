import React, {useContext} from 'react'
import {useForm} from 'react-hook-form'
import useTitle from '../hooks/useTitle';
import UserContext from '../UserContext'
import FormField from '../widgets/FormField';
import NewPasswordFields from '../widgets/NewPasswordFields';
import './ProfileForm.css'

export default function ProfileForm({updateUser, closeModal}) {
  useTitle('User Profile');
  const {user} = useContext(UserContext);
  const {register, handleSubmit, formState: {errors}} = useForm({
    defaultValues: {
      username: user.username,
      imageURL: user.imageURL
    }
  });

  return (
    <form className="ProfileForm" onSubmit={handleSubmit(async data => await updateUser(data))}>
      <h2>Edit Profile</h2>
      <FormField name="username" label="Username" readOnly={true} {...{register, errors}}/>
      <NewPasswordFields editingProfile={true} {...{register, errors}} />
      <FormField name="imageURL" label="Image URL (optional)" {...{register, errors}}/>
      <button type="submit">Save</button>
      <button type="button" onClick={closeModal}>Cancel</button>
      <img src={user.imageURL} alt="" />
    </form>
  );
}
