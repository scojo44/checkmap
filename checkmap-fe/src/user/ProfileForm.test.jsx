import {render, fireEvent, waitFor} from '@testing-library/react'
import {MemoryRouter} from 'react-router-dom'
import UserContext from '../UserContext'
import ProfileForm from './ProfileForm'

const mockUpdateUser = vi.fn();
const context = {
  user: {
    username: 'testuser',
    imageURL: 'u1.jpeg',
    role: 'User',
    lists: []
  }
};

describe('ProfileForm Tests', () => {
  beforeEach(() => {
    mockUpdateUser.mockClear();
  });

  it('Renders without crashing', () => {
    render(
      <MemoryRouter>
        <UserContext.Provider value={context}>
          <ProfileForm />
        </UserContext.Provider>
      </MemoryRouter>
    );
  });

  it('Matches snapshot', () => {
    const {asFragment} = render(
      <MemoryRouter>
        <UserContext.Provider value={context}>
          <ProfileForm />
        </UserContext.Provider>
      </MemoryRouter>
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('Calls the passed function with the submitted values', async () => {
    const {getByText, getByLabelText} = render(
      <MemoryRouter>
        <UserContext.Provider value={context}>
          <ProfileForm updateUser={mockUpdateUser}/>
        </UserContext.Provider>
      </MemoryRouter>
    );

    const imageURL = getByLabelText('Image URL (optional):');
    const password = getByLabelText('New Password (leave blank to not change):');
    const confirm = getByLabelText('Confirm Password:');
    const button = getByText('Save');

    fireEvent.change(imageURL, {target: {value: 'testuser.jpeg'}});
    fireEvent.change(password, {target: {value: 'secret123'}});
    fireEvent.change(confirm, {target: {value: 'secret123'}});
    fireEvent.click(button);

    await waitFor(() => expect(mockUpdateUser).toHaveBeenCalled());
    await waitFor(() => expect(mockUpdateUser).toHaveBeenCalledWith({
      username: 'testuser',
      password: 'secret123',
      confirm: 'secret123',
      imageURL: 'testuser.jpeg'
    }));
  });

  it('Does not call the passed function when confirm does not match password', async () => {
    const {getByText, getByLabelText, container} = render(
      <MemoryRouter>
        <UserContext.Provider value={context}>
          <ProfileForm updateUser={mockUpdateUser}/>
        </UserContext.Provider>
      </MemoryRouter>
    );

    const imageURL = getByLabelText('Image URL (optional):');
    const password = getByLabelText('New Password (leave blank to not change):');
    const confirm = getByLabelText('Confirm Password:');
    const button = getByText('Save');

    fireEvent.change(imageURL, {target: {value: 'testuser.jpeg'}});
    fireEvent.change(password, {target: {value: 'secret123'}});
    fireEvent.change(confirm, {target: {value: 'secret456'}});
    fireEvent.click(button);

    await waitFor(() => expect(mockUpdateUser).not.toHaveBeenCalled());
  });
});
