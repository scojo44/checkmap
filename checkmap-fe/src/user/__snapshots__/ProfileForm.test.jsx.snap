// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ProfileForm Tests > Matches snapshot 1`] = `
<DocumentFragment>
  <form
    class="ProfileForm"
  >
    <h2>
      Edit Profile
    </h2>
    <p
      class="FormField"
    >
      <label
        for="username"
      >
        Username: 
      </label>
      <input
        id="username"
        name="username"
        readonly=""
      />
    </p>
    <p
      class="FormField"
    >
      <label
        for="password"
      >
        New Password (leave blank to not change): 
      </label>
      <input
        autocomplete="new-password"
        id="password"
        name="password"
        type="password"
      />
    </p>
    <p
      class="FormField"
    >
      <label
        for="confirm"
      >
        Confirm Password: 
      </label>
      <input
        id="confirm"
        name="confirm"
        type="password"
      />
    </p>
    <p
      class="FormField"
    >
      <label
        for="imageURL"
      >
        Image URL (optional): 
      </label>
      <input
        id="imageURL"
        name="imageURL"
      />
    </p>
    <button
      type="submit"
    >
      Save
    </button>
    <button
      type="button"
    >
      Cancel
    </button>
    <img
      alt=""
      src="u1.jpeg"
    />
  </form>
</DocumentFragment>
`;
